### PR TITLE
net: ipv4: Implement fallback to LL address 

### LINKED ISSUE #53587: Issue with Auto-IP and Multicast/socket connection
Hello together.

I am using the **STM32H747-DISCO** board with **Zephyr v3.1** .

I configured a socket to receive Multicasts send via UDP by a client software running on a PC and I send back UDP packages to the client software as an answer to the received packages.

The socket is configured in the following way:
```
int32_t init()
{
    int32_t ret;
    struct sockaddr_in addr4;

    addr4.sin_family = AF_INET;
    addr4.sin_port = htons(MULTICAST_LISTEN_PORT);
    addr4.sin_addr.s_addr = INADDR_ANY;

    ret = configSocket((struct sockaddr *)&addr4, sizeof(addr4));
    ...
}

static int32_t configSocket(struct sockaddr *bindAddr, socklen_t bindAddrLen)
{
    struct zsock_addrinfo hints;
    struct zsock_addrinfo* pRetAddrInfo = NULL;
    const struct in_addr* multiAddr = NULL;
    struct net_if_mcast_addr* mcAddr = NULL;
	
    /* Create socket */
    udpData.sock = zsock_socket(bindAddr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
    ...

    /* Bind socket */
    ret = zsock_bind(udpData.sock, bindAddr, bindAddrLen);

    /* Resolve network address */
    memset(&hints, 0, sizeof(struct zsock_addrinfo));
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_DGRAM;
    hints.ai_flags = AI_PASSIVE;
    hints.ai_protocol = IPPROTO_UDP;

    ret = zsock_getaddrinfo(MULTICAST_LISTEN_ADDRESS_STR, MULTICAST_LISTEN_PORT_STR,
                             &hints, &pRetAddrInfo);
    ...

    /* Get IPv4 multicast address */
    multiAddr = &net_sin(pRetAddrInfo->ai_addr)->sin_addr;
    ....

    /* Get network interface */
    struct net_if* ptrNetIf = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));
    ....

    /* Add IPv4 multicast address to network interface */
    mcAddr = net_if_ipv4_maddr_add(ptrNetIf, multiAddr);
    ....
    net_if_ipv4_maddr_join(mcAddr);
    ...
}
```

The socket is polled within a thread to receive the UDP packages and send back answers to the received packages via:
```
void thread()
{
    received = zsock_recvfrom(udpData.sock, udpData.recvBuffer,
			sizeof(udpData.recvBuffer), 0,
			&clientAddr, &clientAddrLen);
    ...
    ret = zsock_sendto(udpData.sock, udpData.sendBuffer, sizeof(struct KistlerDeviceCfg), 0,
		(struct sockaddr *)&serverAddr, serverAddrLen);
    ....
}
```

If a static IP is configured, `zsock_recvfrom` receives the Multicast and `zsock_sendto` sends an answer to the client software.

If the IP settings are changed during running my application from static to DHCP and a DHCP is available `zsock_recvfrom` receives the Multicast and `zsock_sendto` sends an answer to the client software.

If the IP settings are changed during running the application from static to DHCP and a DHCP is not available, Auto-IP is used and Zephyr sets an IP automatically. In that case and `zsock_recvfrom` receives the Multicast but `zsock_sendto` does not send an answer to the client software and returns an error, -EINVAL.

When debugging the issue I found out that in the function `net_if_ipv4_select_src_addr` in line 3215 of zephyr/subsys/net/ip/net_if.c the variable `src` is not set and stays NULL. 

In case of Auto-IP the function call to `net_ipv4_is_ll_addr(dst)` in the if statement in line 3223 returns true as the Auto-IP address is a local-link address. Therefore the else statement in line 3254 is executed. Within that statement src is not set. That's why the if statement in line 3258 is executed. Within this statement `net_if_ipv4_get_global_addr` returns NULL and  `net_ipv4_unspecified_address` returns NULL. That's why at the end `zsock_sendto` fails.  

I do not understand why `src` is not set to a valid value in case of switching from static IP to Auto-IP. Why is `src` not set and `zsock_sendto` fails at the end?

In case of static IP the function call to `net_ipv4_is_ll_addr(dst)` in the if statement in line 3223 returns false and the if statement is executed. In that case line 3227 with the function call `net_if_ipv4_get_best_match` is executed and `src` is set to the static IP address.

### PR DESCRIPTION
**Pull request by rlubos**
According to RFC3927, hosts with only IPv4 LL address should still be
able to send packets to destination w/o IPv4 LL address:

   A host with an IPv4 Link-Local address may send to a destination
   which does not have an IPv4 Link-Local address.  If the host is not
   multi-homed, the procedure is simple and unambiguous: Using ARP and
   forwarding directly to on-link destinations is the default route

This behaviour however was not possible with Zephyr, which only allowed
to use IPv4 LL source address for IPv4 LL destinations.

Fix this, by introducing a final fallback (only if IPv4 autoconf is
enabled), to select IPv4 LL address as a source address if no other
address is available.

Additionally, modify the ARP routine a bit if IPv4 LL address is in use.
There's no really point to forward the packet to gateway if IPv4 LL
address is used, as such addresses are not routable. Instead, try to
find the peer in local network in such case.

Fixes #53587

### CODE REVIEW

File: include/zephyr/net/net_ip.h [hunk 1]
```diff
@@ -730,7 +730,7 @@ static inline bool net_ipv4_is_addr_mcast(const struct in_addr *addr)
  */
 static inline bool net_ipv4_is_ll_addr(const struct in_addr *addr)
 {
-	return (ntohl(UNALIGNED_GET(&addr->s_addr)) & 0xA9FE0000) == 0xA9FE0000;
+	return (ntohl(UNALIGNED_GET(&addr->s_addr)) & 0xFFFF0000) == 0xA9FE0000;
```
Comment (jukkar) [inline include/zephyr/net/net_ip.h:733]: Hmm, what is the actual issue here, why the original check would not work?

Comment (rlubos) [reply include/zephyr/net/net_ip.h:733]: Doing bitwise AND with `0xA9FE0000` could clear the bits (from the subnet part) that were orignally set in the provided address.

Take a IPv4 boradcast address as an example (`255.255.255.255`) - the original approach would modify the subnet part of the address, leading to false match:
`0xFFFFFFFF & 0xA9FE0000` gives `0xA9FE0000`, so the function would indicate it was the LL address. The same issue would apply to any address, that had a match on the bits set in `0xA9FE0000`.

Comment (jukkar) [reply include/zephyr/net/net_ip.h:733]: Indeed, good point.

**Review by jukkar — DISMISSED** [review]

File: subsys/net/l2/ethernet/arp.c [hunk 2]
```diff
@@ -339,15 +339,23 @@ struct net_pkt *net_arp_prepare(struct net_pkt *pkt,
 {
 	struct arp_entry *entry;
 	struct in_addr *addr;
+	bool is_ipv4_ll_used = false;
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/arp.c:342]: just to stay consistent throughout net stack: move the initialized variable at first position in the list.

**Review by tbursztyka — DISMISSED** [review]

File: subsys/net/l2/ethernet/arp.c (hunk 2 above)
Comment (rlubos) [reply subsys/net/l2/ethernet/arp.c]: Done

**Review by jukkar — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**